mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,333 @@
+//! Module for the on-disk data cache implementation.
+
+use std::fs;
+use std::io::{ErrorKind, Read, Write};
+use std::ops::RangeBounds;
+use std::path::PathBuf;
+
+use bytes::Bytes;
+use mountpoint_s3_crt::checksums::crc32c::Crc32c;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use tracing::{error, trace};
+
+use crate::data_cache::DataCacheError;
+
+use super::{BlockIndex, CacheKey, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// Disk and file-layout versioning.
+const CACHE_VERSION: &str = "V1";
+
+/// On-disk implementation of [DataCache].
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+/// Represents a fixed-size chunk of data that can be serialized.
+///
+/// TODO: Add checksum over struct (excl. `data`) to verify block metadata later.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct DataBlock {
+    block_idx: BlockIndex,
+    checksum: u32,
+    data: Bytes,
+    etag: String,
+    s3_key: String,
```

Comment:
It might be nicer to have a `BlockHeader` struct that derives serde and can be checksummed/validated on its own, and then combine that with `data: ChecksummedBytes` here.

Reply:
Oh, I like that! Sounds good.